opt,sql: assorted optimizations

mgartner opened the pull request:
#### opt: add benchmarks for sysbench UPDATE statements

Release note: None

#### execbuilder: add global empty buildScalarCtx

Release note: None

#### sql: refactor MakeUpdater

Some conditionals in the `needsUpdate` have been moved out of the
closure. Also, `needsUpdate` no longer allocates a slice of column IDs.
Instead, it iterates directly over a set of column IDs.

Release note: None

#### opt: convert needsUpdate closure to function

Release note: None

#### sql: lazily allocate includeIndexes

Release note: None

#### sql: combine index entry allocations

Release note: None

#### sql: combine allocations of fetch and update col lists

Release note: None

#### sql: combine index slice allocations for updates and upserts

Release note: None

#### sql/row: lazily compute index column direction slices

The encoding directions of the primary and secondary indexes are only
used in mutations when pretty-printing keys when tracing is enabled and
when `CheckRowSize` emits a log event. These encoding directions are now
lazily computed to eliminate their overhead in the common case.

Release note: None

mgartner committed:
execbuilder: add global empty buildScalarCtx

Release note: None

mgartner committed:
sql: refactor MakeUpdater

Some conditionals in the `needsUpdate` have been moved out of the
closure. Also, `needsUpdate` no longer allocates a slice of column IDs.
Instead, it iterates directly over a set of column IDs.

Release note: None

mgartner committed:
opt: convert needsUpdate closure to function

Release note: None

mgartner committed:
sql: lazily allocate includeIndexes

Release note: None

mgartner committed:
sql: combine index entry allocations

Release note: None

mgartner committed:
sql: combine allocations of fetch and update col lists

Release note: None

mgartner committed:
sql: combine index slice allocations for updates and upserts

Release note: None

mgartner committed:
opt: add benchmarks for sysbench UPDATE statements

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146655)
<!-- Reviewable:end -->

mgartner commented:
Benchmark results here: https://gist.github.com/mgartner/5b4c6c4310471f29f099af91996dafeb

Highlights:

```
name                                                                      old allocs/op  new allocs/op  delta
...
EndToEnd/sysbench-update-index/vectorize=on/Prepared                           284 ± 0%       280 ± 0%  -1.45%  (p=0.000 n=8+8)
EndToEnd/sysbench-update-index/vectorize=off/Simple                            330 ± 0%       326 ± 0%  -1.24%  (p=0.000 n=10+8)
EndToEnd/sysbench-update-index/vectorize=off/Prepared                          272 ± 0%       267 ± 0%  -1.61%  (p=0.000 n=8+8)
EndToEnd/sysbench-update-non-index/vectorize=on/Simple                         348 ± 0%       346 ± 0%  -0.77%  (p=0.000 n=9+8)
EndToEnd/sysbench-update-non-index/vectorize=on/Prepared                       292 ± 0%       289 ± 0%  -1.03%  (p=0.002 n=7+8)
EndToEnd/sysbench-update-non-index/vectorize=off/Simple                        329 ± 0%       326 ± 0%  -0.90%  (p=0.000 n=8+9)
EndToEnd/sysbench-update-non-index/vectorize=off/Prepared                      273 ± 0%       270 ± 0%  -1.17%  (p=0.000 n=8+9)
```

yuzefovich reviewed (approved):
Nice finds! I have a couple of further suggestions, but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-:-OQAyMq3-HpCw1PXQC7n:bkunjz8)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`-- commits` line 36 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-OQAutvtAmgrKIfJzVoz:-OQAutvuDOYulnZonfnJ:b-2532i1):*
nit: we could extend the optimization a bit further to support `returnCols` too - when `returnRows == false`, then the last slice would be empty. We could then apply the same optimization to other mutations. Perhaps leave a TODO for this.
___
*[`pkg/sql/row/helper.go` line 148 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-OQAx-vc9q9d2EJhMSs_:-OQAx-vc9q9d2EJhMSsa:bij617n) ([raw file](https://github.com/cockroachdb/cockroach/blob/028a963e917a6911cec39948bcf3e0815f7c9e07/pkg/sql/row/helper.go#L148)):*
> ```Go
> 
> 	dirs = catalogkeys.AppendIndexKeyValDirs(dirs, rh.TableDesc.GetPrimaryIndex())
> 	rh.primIndexValDirs = dirs[:len(dirs):len(dirs)]
> ```

nit: we can take this one step further and avoid allocating `primIndexValDirs` and `secIndexValDirs` altogether in the common case. These fields are currently used in two scenarios - when we have `traceKV` enabled, and when `CheckRowSize` emits a log event (which happens when we attempt to write a large KV pair).

Perhaps the best way would be to compute both of these fields lazily, the first time either is used. In the common case they won't ever be computed then.


<!-- Sent from Reviewable.io -->

mgartner committed:
sql/row: lazily compute index column direction slices

The encoding directions of the primary and secondary indexes are only
used in mutations when pretty-printing keys when tracing is enabled and
when `CheckRowSize` emits a log event. These encoding directions are now
lazily computed to eliminate their overhead in the common case.

Release note: None

mgartner reviewed (commented):
Great feedback! PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-:-OTd3KgW4HdsYN6TEYQr:b-arlb1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`-- commits` line 36 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-OQAutvtAmgrKIfJzVoz:-OTd2xfw5JrmDCQ6s5_2:bl382xv):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we could extend the optimization a bit further to support `returnCols` too - when `returnRows == false`, then the last slice would be empty. We could then apply the same optimization to other mutations. Perhaps leave a TODO for this.
</blockquote></details>

Good idea. Added a TODO for now.
___
*[`pkg/sql/row/helper.go` line 148 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-OQAx-vc9q9d2EJhMSs_:-OTd3-Zx8fSrQndcn8VK:bpmx73v) ([raw file](https://github.com/cockroachdb/cockroach/blob/028a963e917a6911cec39948bcf3e0815f7c9e07/pkg/sql/row/helper.go#L148)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we can take this one step further and avoid allocating `primIndexValDirs` and `secIndexValDirs` altogether in the common case. These fields are currently used in two scenarios - when we have `traceKV` enabled, and when `CheckRowSize` emits a log event (which happens when we attempt to write a large KV pair).

Perhaps the best way would be to compute both of these fields lazily, the first time either is used. In the common case they won't ever be computed then.
</blockquote></details>

Brilliant idea! Done!


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
Nice! <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 7 files at r10, 3 of 3 files at r11, 1 of 1 files at r12, 1 of 1 files at r13, 1 of 1 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, 1 of 1 files at r17, 5 of 5 files at r18, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146655#-:-OTdTUzODhc5KZSpXe85:be8rotp)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner commented:
Re-ran the same benchmarks: https://gist.github.com/mgartner/268ef9fea4c17378a287573ff971d9b8

Managed to shave off some more allocations thanks to Yahor's suggestions!

```
EndToEnd/sysbench-update-index/vectorize=on/Simple                             343 ± 0%       337 ± 0%   -1.86%  (p=0.000 n=12+13)
EndToEnd/sysbench-update-index/vectorize=on/Prepared                           285 ± 0%       279 ± 0%   -2.20%  (p=0.000 n=13+12)
EndToEnd/sysbench-update-index/vectorize=off/Simple                            330 ± 1%       325 ± 1%   -1.66%  (p=0.000 n=13+12)
EndToEnd/sysbench-update-index/vectorize=off/Prepared                          272 ± 0%       266 ± 0%   -2.18%  (p=0.000 n=13+11)
EndToEnd/sysbench-update-non-index/vectorize=on/Simple                         348 ± 0%       344 ± 1%   -1.03%  (p=0.000 n=12+13)
EndToEnd/sysbench-update-non-index/vectorize=on/Prepared                       292 ± 1%       288 ± 1%   -1.28%  (p=0.000 n=12+13)
EndToEnd/sysbench-update-non-index/vectorize=off/Simple                        328 ± 0%       324 ± 0%   -1.34%  (p=0.000 n=12+11)
EndToEnd/sysbench-update-non-index/vectorize=off/Prepared                      272 ± 0%       268 ± 1%   -1.52%  (p=0.000 n=14+13)
```

mgartner commented:
TFTR!

bors r+
